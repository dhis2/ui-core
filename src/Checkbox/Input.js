import React, { Component, createRef } from 'react'
import propTypes from '@dhis2/prop-types'

export class Input extends Component {
    ref = createRef()

    componentDidMount() {
        if (this.props.focus) {
            this.ref.current.focus()
        }
    }

    render() {
        const {
            onChange,
            checked,
            disabled,
            name,
            value,
            onFocus,
            onBlur,
            tabIndex,
        } = this.props

        return (
            <div>
                <input
                    ref={this.ref}
                    type="checkbox"
                    name={name}
                    value={value}
                    onChange={onChange}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    checked={checked}
                    disabled={disabled}
                    tabIndex={tabIndex}
                />

                <style jsx>{`
                    div {
                        width: 0;
                        height: 0;
                        overflow: hidden;
                    }
                `}</style>
            </div>
        )
    }
}

Input.propTypes = {
    onChange: propTypes.func.isRequired,
    name: propTypes.string.isRequired,
    value: propTypes.string,

    tabIndex: propTypes.string,

    onFocus: propTypes.func,
    onBlur: propTypes.func,
    checked: propTypes.bool,
    disabled: propTypes.bool,
    focus: propTypes.bool,
}
