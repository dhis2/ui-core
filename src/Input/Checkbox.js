import React from 'react';
import PropTypes from 'prop-types';
import './checkbox.css';
import { bemClassNames } from '../utils';

const bem = bemClassNames('d2ui-checkbox');

const Checkbox = props => {
    return (
        <div class="mdc-checkbox">
            <input type="checkbox" class="mdc-checkbox__native-control" id="checkbox-1" />
            <div class="mdc-checkbox__background">
                <svg class="mdc-checkbox__checkmark" viewBox="0 0 24 24">
                    <path
                        class="mdc-checkbox__checkmark-path"
                        fill="none"
                        d="M1.73,12.91 8.1,19.28 22.79,4.59"
                    />
                </svg>
                <div class="mdc-checkbox__mixedmark" />
            </div>
        </div>
    );
};

Checkbox.propTypes = {};

export default Checkbox;
