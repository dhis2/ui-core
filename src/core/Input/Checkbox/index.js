/** @format */

import React from 'react'
import PropTypes from 'prop-types'
import './styles.css'
import { bemClassNames, getRequiredText } from '../../../utils'
import ToggleField from '../ToggleField'
import FieldWrap from '../FieldWrap'

const bem = bemClassNames('checkbox')

const Checkbox = ({
    label,
    checked,
    valid,
    warning,
    error,
    disabled,
    onChange,
    dense,
    block,
    helpText,
    required,
}) => (
    <FieldWrap {...{ valid, warning, disabled, error, dense, block, helpText }}>
        <ToggleField disabled={disabled} dense={dense}>
            <div className={bem.b({ disabled })}>
                <input
                    type="checkbox"
                    className={bem.e('native-control')}
                    checked={checked}
                    disabled={disabled}
                    onChange={onChange}
                />
                <div className={bem.e('background')}>
                    <svg className={bem.e('checkmark')} viewBox="0 0 24 24">
                        <path
                            className={bem.e('checkmark-path')}
                            fill="none"
                            d="M1.73,12.91 8.1,19.28 22.79,4.59"
                        />
                    </svg>
                    <div className={bem.e('mixedmark')} />
                </div>
            </div>
            <span className={bem.e('label-text', { disabled })}>
                {getRequiredText(label, required)}
            </span>
        </ToggleField>
    </FieldWrap>
)

Checkbox.propTypes = {
    label: PropTypes.string,
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    valid: PropTypes.bool,
    warning: PropTypes.bool,
    error: PropTypes.bool,
    dense: PropTypes.bool,
    helpText: PropTypes.string,
    block: PropTypes.bool,
    required: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
}

Checkbox.defaultProps = {
    value: false,
    disabled: false,
}

export { Checkbox }
export default Checkbox
