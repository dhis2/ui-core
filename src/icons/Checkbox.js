import React from 'react'
import propTypes from '@dhis2/prop-types'

export function Unchecked({ className }) {
    return (
        <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 48 48"
            className={className}
        >
            <path d="M38 10v28H10V10h28m0-4H10c-2.21 0-4 1.79-4 4v28c0 2.21 1.79 4 4 4h28c2.21 0 4-1.79 4-4V10c0-2.21-1.79-4-4-4z" />
            <style jsx>{`
                svg {
                    fill: inherit;
                    height: 24px;
                    width: 24px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            `}</style>
        </svg>
    )
}
Unchecked.propTypes = {
    className: propTypes.string,
}

export function Checked({ className }) {
    return (
        <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 48 48"
            className={className}
        >
            <path d="M38 6H10c-2.21 0-4 1.79-4 4v28c0 2.21 1.79 4 4 4h28c2.21 0 4-1.79 4-4V10c0-2.21-1.79-4-4-4zM20 34L10 24l2.83-2.83L20 28.34l15.17-15.17L38 16 20 34z" />
            <style jsx>{`
                svg {
                    fill: inherit;
                    height: 24px;
                    width: 24px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            `}</style>
        </svg>
    )
}
Checked.propTypes = {
    className: propTypes.string,
}

export function Indeterminate({ className }) {
    return (
        <svg
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 24 24"
            className={className}
        >
            <path d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z" />
            <style jsx>{`
                svg {
                    fill: inherit;
                    height: 24px;
                    width: 24px;
                    vertical-align: middle;
                    pointer-events: none;
                }
            `}</style>
        </svg>
    )
}
Indeterminate.propTypes = {
    className: propTypes.string,
}
