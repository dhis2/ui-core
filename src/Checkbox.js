import cx from 'classnames'
import propTypes from 'prop-types'
import React, { Component, Fragment } from 'react'

import { statusPropType } from './common-prop-types.js'

import { theme } from './theme.js'

import { Icon } from './Checkbox/Icon.js'
import { Input } from './Checkbox/Input.js'
import { Label } from './Checkbox/Label.js'

/**
 * @module
 * @param {Checkbox.PropTypes} props
 * @returns {React.Component}
 * @example import { Checkbox } from @dhis2/ui-core
 * @see Specification: {@link https://github.com/dhis2/design-system/blob/master/atoms/checkbox.md|Design system}
 * @see Live demo: {@link /demo/?path=/story/checkbox--default|Storybook}
 */
class Checkbox extends Component {
    constructor(props) {
        super(props)

        this.state = {
            focus: props.initialFocus,
        }
    }

    onFocus = e => {
        this.setState({ focus: true })

        if (this.props.onFocus) {
            this.props.onFocus(e)
        }
    }

    onBlur = e => {
        this.setState({ focus: false })

        if (this.props.onBlur) {
            this.props.onBlur(e)
        }
    }

    render() {
        const {
            checked = false,
            className,
            disabled,
            error,
            icon,
            indeterminate,
            label,
            name,
            onChange,
            required,
            tabIndex,
            valid,
            value,
            warning,
        } = this.props
        const { focus } = this.state

        return (
            <Fragment>
                <label
                    className={cx('base', className, {
                        disabled,
                        focus,
                    })}
                >
                    <Input
                        name={name}
                        value={value}
                        checked={checked}
                        disabled={disabled}
                        focus={focus}
                        onChange={onChange}
                        onFocus={this.onFocus}
                        onBlur={this.onBlur}
                        tabIndex={tabIndex}
                    />

                    <Icon
                        focus={focus}
                        checked={checked}
                        disabled={disabled}
                        valid={valid}
                        error={error}
                        warning={warning}
                        indeterminate={indeterminate}
                    />

                    {icon}

                    <Label required={required} disabled={disabled}>
                        {label}
                    </Label>

                    <style jsx>{`
                        label {
                            display: flex;
                            flex-direction: row;
                            align-items: center;
                            justify-content: flex-start;
                            cursor: pointer;
                            pointer-events: all;
                            user-select: none;
                        }

                        .disabled {
                            cursor: not-allowed;
                            color: ${theme.disabled};
                        }

                        .focus {
                            outline: 1px;
                        }
                    `}</style>
                </label>
            </Fragment>
        )
    }
}

/**
 * @typedef {Object} PropTypes
 * @static
 *
 * @prop {function} onChange - The handler which is triggered when the
 * component changes
 * @prop {string} value
 * @prop {string} name
 * @prop {string} label
 *
 * @prop {string} [tabIndex]
 * @prop {string} [className]
 * @prop {function} [onFocus]
 * @prop {function} [onBlur]
 *
 * @prop {Element} [icon]
 * @prop {boolean} [indeterminate]
 * @prop {boolean} [required]
 * @prop {boolean} [checked]
 * @prop {boolean} [disabled]
 * @prop {boolean} [initialFocus]
 * @prop {boolean} [valid] - `valid`, `warning`, and `error`, are mutually exclusive
 * @prop {boolean} [warning]
 * @prop {boolean} [error]
 */
Checkbox.propTypes = {
    onChange: propTypes.func.isRequired,

    value: propTypes.string,
    name: propTypes.string.isRequired,
    label: propTypes.string.isRequired,
    tabIndex: propTypes.string,
    className: propTypes.string,

    onFocus: propTypes.func,
    onBlur: propTypes.func,

    icon: propTypes.element,

    indeterminate: propTypes.bool,
    required: propTypes.bool,
    checked: propTypes.bool,
    disabled: propTypes.bool,
    valid: statusPropType,
    warning: statusPropType,
    error: statusPropType,
    initialFocus: propTypes.bool,
}

export { Checkbox }
