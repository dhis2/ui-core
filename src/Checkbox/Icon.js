import React, { Fragment } from 'react'
import propTypes from 'prop-types'
import cx from 'classnames'
import { resolve } from 'styled-jsx/css'

import { colors, theme } from '../theme'
import { Indeterminate, Checked, Unchecked } from '../icons/Checkbox'

const icons = resolve`
    svg {
        height: 24px;
        width: 24px;
        fill: ${theme.default};
    }

    .checked {
        fill: ${colors.teal400};
    }

    .disabled {
        fill: ${theme.disabled};
    }

    .error {
        fill: ${theme.error};
    }

    .valid {
        fill: ${theme.valid};
    }

    .warning {
        fill: ${theme.warning};
    }
`

export const Icon = ({
    checked,
    disabled,
    valid,
    error,
    warning,
    indeterminate,
}) => {
    const classes = cx(icons.className, {
        checked: checked && !valid && !error && !warning,
        disabled,
        valid,
        error,
        warning,
    })

    return (
        <Fragment>
            {indeterminate ? (
                <Indeterminate className={classes} />
            ) : checked ? (
                <Checked className={classes} />
            ) : (
                <Unchecked className={classes} />
            )}

            {icons.styles}
        </Fragment>
    )
}

Icon.propTypes = {
    checked: propTypes.bool,
    disabled: propTypes.bool,
    valid: propTypes.bool,
    error: propTypes.bool,
    warning: propTypes.bool,
    indeterminate: propTypes.bool,
}
