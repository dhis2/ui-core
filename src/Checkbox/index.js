import React, { Fragment } from 'react'
import propTypes from 'prop-types'
import cx from 'classnames'
import css from 'styled-jsx/css'

import { colors, theme } from '../theme'
import { Indeterminate, Checked, Unchecked } from '../icons/Checkbox'
import { Icon } from './Icon'
import { Label } from './Label'
import { Input } from './Input'

class Checkbox extends React.Component {
    state = {
        indeterminate: this.props.indeterminate,
        checked: this.props.defaultChecked || false,
    }

    onChange = evt => {
        const { checked } = evt.target
        const indeterminate = this.state.indeterminate ? false : null

        this.setState({
            checked,
            indeterminate,
        })

        this.props.onChange(this.props.name, checked)
    }

    render() {
        const {
            required,
            className,
            disabled,
            valid,
            error,
            warning,
            checked = this.state.checked,
        } = this.props

        return (
            <label
                className={cx('base', className, {
                    disabled: this.props.disabled,
                })}
            >
                <Input
                    onChange={this.onChange}
                    checked={checked}
                    disabled={this.props.disabled}
                />

                <Icon
                    checked={checked}
                    disabled={disabled}
                    valid={valid}
                    error={error}
                    warning={warning}
                    indeterminate={this.state.indeterminate}
                />

                <Label required={required}>{this.props.label}</Label>

                <style jsx>{`
                    label {
                        display: flex;
                        flex-direction: row;
                        align-items: center;
                        justify-content: flex-start;
                        cursor: pointer;
                        pointer-events: all;
                        user-select: none;
                    }

                    .disabled {
                        cursor: not-allowed;
                        color: ${theme.disabled};
                    }
                `}</style>
            </label>
        )
    }
}

Checkbox.propTypes = {
    name: propTypes.string.isRequired,

    className: propTypes.string,
    label: propTypes.string,

    indeterminate: propTypes.bool,
    required: propTypes.bool,
    checked: propTypes.bool,
    defaultChecked: propTypes.bool,
    disabled: propTypes.bool,
    valid: propTypes.bool,
    warning: propTypes.bool,
    error: propTypes.bool,

    onChange: propTypes.func,
}

export { Checkbox }
