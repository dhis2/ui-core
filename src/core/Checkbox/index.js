import React from 'react'
import PropTypes from 'prop-types'

import Icon from '../Icon'
import cx, { rx } from './styles'

class Checkbox extends React.Component {
    state = {
        indeterminate: this.props.indeterminate,
    }

    onChange = () => {
        this.props.onChange(this.props.name, !this.props.checked)

        if (this.state.indeterminate) {
            this.setState({ indeterminate: false })
        }
    }

    render() {
        const { required, status } = this.props
        let name = 'check_box_outline_blank'

        if (this.state.indeterminate) {
            name = 'indeterminate_check_box'
        } else if (this.props.checked) {
            name = 'check_box'
        }

        const icon = (
            <Icon
                name={name}
                className={cx(`${status}-icon`, {
                    disabled: this.props.disabled,
                    checked: this.props.checked,
                })}
            />
        )

        return (
            <label
                className={rx('base', {
                    disabled: this.props.disabled,
                })}
            >
                <input
                    type="checkbox"
                    onChange={this.onChange}
                    checked={this.props.checked}
                    disabled={this.props.disabled}
                />
                {icon}
                <span className={rx('label', { required })}>
                    {this.props.label}
                </span>
            </label>
        )
    }
}

Checkbox.defaultProps = {
    label: '',
    checked: false,
    required: false,
    disabled: false,
    indeterminate: false,
    status: 'default',
}

Checkbox.propTypes = {
    name: PropTypes.string.isRequired,
    label: PropTypes.string,
    checked: PropTypes.bool,
    required: PropTypes.bool,
    disabled: PropTypes.bool,
    indeterminate: PropTypes.bool,
    status: PropTypes.oneOf(['default', 'valid', 'warning', 'error']),
    onChange: PropTypes.func.isRequired,
}

export { Checkbox }
export default Checkbox
