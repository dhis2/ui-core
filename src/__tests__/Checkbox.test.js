import React from 'react'
import { shallow, mount } from 'enzyme'
import { Checkbox } from '../Checkbox.js'

const defaultProps = {
    name: 'name',
    label: 'label',
    onChange: () => {},
}

describe('<Checkbox>', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Checkbox {...defaultProps} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('calls onChange when changed', () => {
        const onChange = jest.fn()
        const wrapper = mount(
            <Checkbox {...defaultProps} onChange={onChange} />
        )

        wrapper.find('input').simulate('change', { target: { checked: true } })

        expect(onChange).toHaveBeenCalled()
    })
})
