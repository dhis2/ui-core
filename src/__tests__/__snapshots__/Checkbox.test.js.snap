// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> renders correctly 1`] = `
<Fragment>
  <label
    className="jsx-1861485737 base"
  >
    <Input
      checked={false}
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
    />
    <Icon
      checked={false}
    />
    <Label>
      label
    </Label>
    <JSXStyle
      dynamic={
        Array [
          "#6e7a8a",
        ]
      }
      id="261820537"
    >
      label.__jsx-style-dynamic-selector{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:start;-webkit-justify-content:flex-start;-ms-flex-pack:start;justify-content:flex-start;cursor:pointer;pointer-events:all;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;}.disabled.__jsx-style-dynamic-selector{cursor:not-allowed;color:#6e7a8a;}.focus.__jsx-style-dynamic-selector{outline:1px;}
    </JSXStyle>
  </label>
</Fragment>
`;
